### PR TITLE

SLCORE-786 allow blame with given content for unsaved files

### PR DESCRIPTION
**Pull request by serhat-yenican-sonarsource**
# For SonarSourcers:

- [ ] Prefix the commit message with the ticket number, i.e. `SLCORE-XXXX` (or use `NO-JIRA` for a trivial change not tracked by a ticket)
- [ ] When changing an API:
    - [ ] Explain in the JavaDoc the purpose of the new API
    - [ ] Document the change in [API_CHANGES.md](https://github.com/SonarSource/sonarlint-core/blob/master/API_CHANGES.md)
    - [ ] If the change breaks the current API, explicitly communicate those to the impacted consumers prior to merging (eg. IDE squad)
- [ ] Make sure the tests adhere to the convention:
    - [ ] All test method names should use `snake_case`, for example: `test_validate_input`.
- [ ] Make sure checks are green: build passes, Quality Gate is green

# For external contributors:

In addition to the above, please review our [contribution guidelines](https://github.com/SonarSource/sonarlint-core/blob/master/docs/contributing.md) and ensure your pull request adheres to the following guidelines:

- [ ] Please explain your motives to contribute this change: what problem you are trying to fix, what improvement you are trying to make
- [ ] Use the following formatting style: [SonarSource/sonar-developer-toolset](https://github.com/SonarSource/sonar-developer-toolset#code-style)
- [ ] Provide a unit test for any code you changed

### CODE REVIEW

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/analysis/AnalysisStartedEvent.java [hunk 1]
```diff
@@ -62,4 +64,21 @@ public Set<URI> getFileUris() {
   public boolean isTrackingEnabled() {
     return enableTracking;
   }
+
+  public UnaryOperator<String> getFileContentProvider() {
+    return path -> files.stream()
```
Comment (vojtech-suchy-sonarsource) [inline backend/core/src/main/java/org/sonarsource/sonarlint/core/analysis/AnalysisStartedEvent.java:69]: Do we need to take into account that the  filePath we set via the [RepositoryBlameCommand.seFilePath() will always have the UNIX format](https://github.com/SonarSource/sonarlint-core/blob/8dc549e82f5674928a39f2ea806e4e693a685f87/backend/commons/src/main/java/org/sonarsource/sonarlint/core/commons/GitBlameUtils.java#L79)? Do any of the paths here also need to be in UNIX format?

File: backend/commons/src/main/java/org/sonarsource/sonarlint/core/commons/GitBlameUtils.java [hunk 2]
```diff
@@ -67,12 +71,14 @@ private static RepositoryBuilder getVerifiedRepositoryBuilder(Path basedir) {
     return builder;
   }
 
-  private static SonarLintBlameResult blameWithFilesGitCommand(Repository repo, Set<Path> gitRelativePath) {
+  private static SonarLintBlameResult blameWithFilesGitCommand(Repository repo, Set<Path> gitRelativePath, @Nullable UnaryOperator<String> fileContentProvider) {
```
Comment (vojtech-suchy-sonarsource) [inline backend/commons/src/main/java/org/sonarsource/sonarlint/core/commons/GitBlameUtils.java:74]: This is a minor nit-pick - I usually prefer having two separate constructors so we don't need to update the existing code by passing `nulls` (such as in the tests)

**Review by vojtech-suchy-sonarsource — COMMENTED** [review]
Over all looks good! Thanks for looking into this.

Mainly had one question on whether we need to convert any paths to be unix style (see the comment below)

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/analysis/AnalysisStartedEvent.java (hunk 1 above)
Comment (serhat-yenican-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/analysis/AnalysisStartedEvent.java:69]: That's a very good catch! Yes, we should definitely take this into account. Thanks.

Comment (serhat-yenican-sonarsource) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/analysis/AnalysisStartedEvent.java:69]: I modified the `it_should_blame_with_given_contents` test to contain inner path logix to be able to test the forward-back slash scenario.

**Review by vojtech-suchy-sonarsource — APPROVED** [review]
Splendid 👏

**Pull request merged**